design-documents/buildattr64-rationale.rst

```diff
@@ -0,0 +1,540 @@
+  process. For example if all tags in a subsection have the value 0
+  or 1, and have a partial order that is identical or reverse of the
+  tags arithmetic value, then an implementation can represent this as
+  a bit vector and use the binary ``or`` or ``and`` operator to do the
+  join.
+
+Alternative one .aeabi_attributes section
+-----------------------------------------
+
+Instead of grouping attributes with similar properties into their own
+subsection, we follow (ADDENDA32_) and have one single public
+.aeabi_attributes section that contains all the public attributes. Or
+have one public .aeabi_attributes section for ULEB128 and one for
+NTBS.
+
+This would be simpler to encode in assembly, with only one directive
+required, but would make the specification harder to understand
+and maintain consistency over time.
+
+All tags ULEB128 or NTBS
+========================
+
+A build attributes consumer can more easily skip tags it doesn't
+understand if the encoding for the value is known ahead of time. The
+rule in (ADDENDA32_) that used whether the tag was even or odd was
+error prone as the rule is not widely known. Moreover almost all
+tags are ULEB128 so the even odd rule wastes almost half the encoding
+space for tags.
+
+Reference Partial Order
+=======================
+
+A majority of build attributes have the partial order either matching
+or reversing the build attribute arithmetic order. As (ADDENDA32_)
+states it is usually obvious when this is the case. However there are
+more complex cases and in these cases it can help to state what the
+reference intention is for the tag.
+
+Appendix: Recommended Tool Interface for aeabi subsections
+==========================================================
+
+Compiler
+--------
+
+Build attributes are set by the compiler based on command-line
+options.  For example the clang and gcc ``-mbranch-protection`` option
+can be used to derive ``Tag_Feature_BTI`` and ``Tag_Feature_PAC`` and
+``Tag_Feature_GCS``.
+
+Individual functions can be given different values from the file scope
+command-line options. The file scope build attributes should still be
+derived from the file scope command-line options, or module level encodings
+of the file scope command-line options in the case of link time optimization.
+It is the user's responsibility that the individual functions are used in a
+compatible way to the file scope build attributes.
+
+Assembler
+---------
+
+Where possible the assembler can derive build attributes from the
+assembler's command line options in the same way as the compiler.  For
+options that cannot be derived the following directives can be used to
+construct "aeabi" prefixed subsections.
+
+::
+   .aeabi_subsection name [optional] [, parameter type]
+
+
+*name*
+
+Create or switch the current subsection to *name*.
+
+*optional*
+
+This field is optional and only applies to subsection names with a
+prefix of "aeabi".  The default value is 1 for optional.
```

Comment:
Starting to doubt whether it is a good idea to have optional parameters with default values.
It creates the possibility for user mistakes when writing build attributes as assembly and implementation mistakes. 
Are there any real benefits for allowing the omission of parameters?

Reply:
Default values are necessary for reasoning about the properties of old object files in new toolchains and vice-versa. If we introduce a new attribute in ABI release N, then toolchains that implement some version prior to N won't have that attribute, toolchains need to be able to reason about that case as we can't guarantee that everyone recompiles all their objects, including all their dependencies, all of the time. An explicit default value of 0 forces the attribute definition to think about backwards compatibility. With the most common kinds of attribute combination  being "all objects must have property X to set X in output", "If at least one object has property X to set X in output" or "all objects must have the same property for the combination to be valid" then 0 works well.

Reply:
Thank you for the reply, it is a bit difficult to understand.

Given old object file O1 which does not have attrX,
Being handled with new object files O2, O3 that has attrX.
The way to handle O1, O2 and O3 together will be implemented in the logic of the handling tool (e.g. linker)
Something like "if no attrx, then..."

It is not so clear how having default values for attrX helps this process.
1) Having default value for attrX can not imply that old object has to be treated like it has the default value, because there is no way to know if it will make sense for future attributes.
2)  "An explicit default value of 0 forces the attribute definition to think about backwards compatibility" - Why? when there is an old object involved, it is not the same as having new object with no explicit attribute (see (1)) It is a completely different case that has to be handled anyhow. e.g. even in a world without old object files, there could be a case of a missing attribute, either buy design (if the attribute is optional) or mistake (if it is mandatory)

It seems to be completely unrelated?

Reply:
There is some logic to the idea that the default is '0', and it means that no attribute -> value is '0', and then it imply that all other objects can not have '1'.
However, if this is the intention, do we really want to force all future developers to comply with this logic? It is a bit micromanagy on our part, and it might be better to leave those decisions to the future implementers.

Reply:
Sorry, I am gathering my thoughts slowly. Just want to add this:
Handling old object files has to be done anyhow, and I believe that having default values will hardly effect it or will hint anything to those who will implement it, while it will add the need to remember that 1) there are default values and 2) what they are to everyone who read/write assembly, or handle the tools that parse/write assembly/elf. We adding complexity and making things more error prone for benefits that does not seems quite there.

Reply:
(sorry, a very last comment ) It will be clear from the partial order defined on the attribute, what should be the way to handle the case of missing attribute. (or "what should be the default value for missing attribute") and in the process of implementation and review, it should be done correctly, is there a need to hint it by setting a default value?

Reply:
Hope this goes through all the questions:

A given tool will understand a non-strict subset of the documented attributes and documented subsections of the current specification.

A build attributes producer like an assembler or compiler can only set the attributes and subsections it knows about. A build attributes consumer like a linker can only take action  (the encoding permits it to read the attributes even if it doesn't understand them) on attributes and subsections it knows about.

On the producer side, the attributes and subsections that the tool doesn't know about will not be present. When the intended value of the attribute is the same as the default then it is up to the producer whether to emit the attribute or leave it out. Personally I would expect tools like compilers to just set the attribute explicitly to 0, however in hand-written assembly I would expect humans to only set the ones they care about. It isn't realistic to expect a human to look up all the attributes in the spec.

On the consumer side it is clear what happens if an attribute is encountered that the consumer doesn't understand. If it is in an optional subsection, it ignores it, if it is an non-optional subsection it gives a diagnostic.

If the consumer encounters an object file that is missing an attribute that it understands it uses the default 0 value for the object. As you suggest the partial order and combination rule will tell the consumer how to update the output value of the attribute.

Reply:
Thanks, I was confusing Assembly default values, such as subsection header values such as optional/value and ELF default value such as an entire tag.
ELF does not have default values like Assembly, therefore my point was not relevant.